File: Erlang/funzioni.erl
-module(funzioni).
-compile(export_all).
%include la libreria per poter effettuare le query
-include_lib("stdlib/include/qlc.hrl").
%include il file blogchat_DB.hrl
-include("blogchat_DB.hrl").
%include il file sessione.hrl
-include("sessione.hrl").
%include il file yaws_api.hrl
-include("C:/Program Files (x86)/Yaws-1.98/include/yaws_api.hrl").

%funzioni.erl:file contenente le funzioni utilizzate all'interno di BlogChat


%funzione che esegue una query	
do(Q) ->
    F = fun() -> qlc:e(Q) end,
    {atomic, Val} = mnesia:transaction(F),
    Val.
	
%funzione che controlla se un utente è già stato inserito all'interno del DB
controlla_utente(Nome_utente) ->
	 do(qlc:q([X#utenti.nome_utente || X <- mnesia:table(utenti),
			     X#utenti.nome_utente =:= Nome_utente
				])).
				
%funziona che registra un nuovo utente all'interno del DB
registra_utente(Nome,Cognome,Nascita,Nome_utente,Email,Pass) -> 
	mnesia:dirty_update_counter(appoggio,numUtenti,+1),
	ListaId=leggi_id(numUtenti),
	[Id | Lista ] = ListaId,
	Row = #utenti{id=Id, nome=Nome, cognome=Cognome, nascita=Nascita, nome_utente=Nome_utente, email=Email, pass=Pass},
    F = fun() ->
		mnesia:write(Row)
	end,
    mnesia:transaction(F).


%legge l'id corrente dalla tabella appoggio del label Target per salvare utenti, post o commenti
leggi_id(Target)->
	do(qlc:q([X#appoggio.numero || X <- mnesia:table(appoggio),
			     X#appoggio.label =:= Target
				])).

%recupera la password dell'utente che sta effettuando il login
recupera_password(Nome_Utente) ->
	do(qlc:q([X#utenti.pass || X <- mnesia:table(utenti),
			     X#utenti.nome_utente =:= Nome_Utente
				])).

%recupera le informazioni dell'utente per visualizzarle nella "bacheca"
recuperaInformazioni(Nome_Utente) ->
	do(qlc:q([{X#utenti.nome, X#utenti.cognome, X#utenti.nascita, X#utenti.email} || X <- mnesia:table(utenti),
			     X#utenti.nome_utente =:= Nome_Utente
				])).
				
%funzione che restituisce una lista vuota nel caso in cui l'avatar non è associato ancora all'utente, altrimenti restituisce una lista con il nome dell'immagine caricata
caricamentoAvatar(Nome_utente) ->
	 do(qlc:q([X#avatar.nome_immagine || X <- mnesia:table(avatar),
				Y <- mnesia:table(utenti),
				X#avatar.id_utente =:= Y#utenti.id,
			    Y#utenti.nome_utente =:= Nome_utente
				])).
				
%funzione che controlla se il tipo di un file caricato è corretto oppure no
controllaTipo(TipoControllo) ->
	if
		TipoControllo =:= "image/jpeg"; TipoControllo =:= "image/gif" ; TipoControllo =:= "image/png" -> true;
		true -> false
	end.

%funzione che registra all'interno del DB il nome dell'immagine personale dell'utente
registraAvatar(NomeUtente,NomeFile,TemporaryFile) ->
	Val = do(qlc:q([{X#utenti.id} || X <- mnesia:table(utenti),
				X#utenti.nome_utente =:= NomeUtente])),
	[{IdUtente}] = Val,
	Row = #avatar{id_utente=IdUtente, nome_immagine=NomeFile},
	Avatar = caricamentoAvatar(NomeUtente),
			if
				length(Avatar) == 0 -> F = fun() ->
												mnesia:write(Row)
												end,
											    mnesia:transaction(F),
											    ImmagineNuova = string:concat("avatar/",NomeFile),
												file:copy(TemporaryFile,ImmagineNuova);
				length(Avatar) > 0 ->  Val2 = do(qlc:q([{X#avatar.nome_immagine} || X <- mnesia:table(avatar),
											X#avatar.id_utente =:= IdUtente])),
											[ImmagineDaCanc] = Val2,
											ImmagineVecchia = string:concat("avatar/",ImmagineDaCanc),
										   	file:delete(ImmagineVecchia),
									G = fun() ->
												mnesia:delete({avatar,IdUtente})
												end,
										   	 	mnesia:transaction(G),
									H = fun() ->
												mnesia:write(Row)
												end,
												mnesia:transaction(H),
												ImmagineNuova2 = string:concat("avatar/",NomeFile),
												file:copy(TemporaryFile,ImmagineNuova2)
			end.
%funzione registra_post (NomeUtente,Titolo,Contenuto): registra il post il cui titolo e il contenuto sono passati come argomenti, aggiungendo la data corrente e l'identificatore del post
registra_post(NomeUtente,Titolo,Contenuto) ->
			mnesia:dirty_update_counter(appoggio,numPost,+1),
			ListaId=leggi_id(numPost),
			[Id | Lista ] = ListaId,
			Val = do(qlc:q([{X#utenti.id} || X <- mnesia:table(utenti),
				X#utenti.nome_utente =:= NomeUtente])),
			[{IdUtente}] = Val,
			DataPost = erlang:localtime(),
			Row = #post{id_post=Id, id_utente=IdUtente, nome_post=Titolo, contenuto_post=Contenuto, data=DataPost},
   			F = fun() ->
						mnesia:write(Row)
						end,
   			mnesia:transaction(F).

%funzione che legge i post di un singolo utente nella home page
leggi_post_utente(NomeUtente)->
		QH = do(qlc:q([{X#post.id_post,X#post.data,X#post.nome_post,X#post.contenuto_post,Y#utenti.nome,Y#utenti.cognome} || X <- mnesia:table(post),
						Y <- mnesia:table(utenti),
						X#post.id_utente =:= Y#utenti.id,
						Y#utenti.nome_utente =:= NomeUtente])),
		QL =qlc:sort(QH,{order,descending}),
		QC = qlc:cursor(QL),
		Post = qlc:next_answers(QC,5),
		qlc:delete_cursor(QC),
		Post.

%funzione che legge i post meno recenti di un singolo utente (per la sezione archivio)
leggi_postVecchi_utente(NomeUtente,NumeroPagina) ->
		if
			NumeroPagina == 1 ->
						QH = do(qlc:q([{X#post.id_post,X#post.data,X#post.nome_post,X#post.contenuto_post,Y#utenti.nome,Y#utenti.cognome} || X <- mnesia:table(post),
										Y <- mnesia:table(utenti),
										X#post.id_utente =:= Y#utenti.id,
										Y#utenti.nome_utente =:= NomeUtente])),
						QL =qlc:sort(QH,{order,descending}),
						QC = qlc:cursor(QL),
						PostNuovi = qlc:next_answers(QC,5),
						Post = qlc:next_answers(QC,5),
						qlc:delete_cursor(QC),
						Post;
				true -> 
						QH = do(qlc:q([{X#post.id_post,X#post.data,X#post.nome_post,X#post.contenuto_post,Y#utenti.nome,Y#utenti.cognome} || X <- mnesia:table(post),
										Y <- mnesia:table(utenti),
										X#post.id_utente =:= Y#utenti.id,
										Y#utenti.nome_utente =:= NomeUtente])),
						QL =qlc:sort(QH,{order,descending}),
						QC = qlc:cursor(QL),
						PostNuovi = qlc:next_answers(QC,5),
						Numero = (NumeroPagina-1)*5,
						PostNonConsiderati = qlc:next_answers(QC,Numero),
						Post = qlc:next_answers(QC,5),
						qlc:delete_cursor(QC),
						Post
		end.



%funzione che legge i post di tutti gli utenti (per la sezione notizie)
leggi_post_utenti(NumeroPagina) ->
		if
			NumeroPagina == 1 ->QH = do(qlc:q([{X#post.id_post,X#post.data,X#post.nome_post,X#post.contenuto_post,Y#utenti.nome,Y#utenti.cognome} || X <- mnesia:table(post),
						Y <- mnesia:table(utenti),
						X#post.id_utente =:= Y#utenti.id])),
						QL =qlc:sort(QH,{order,descending}),
						QC = qlc:cursor(QL),
						Post = qlc:next_answers(QC,20),
						qlc:delete_cursor(QC),
						Post;
			true -> QH = do(qlc:q([{X#post.id_post,X#post.data,X#post.nome_post,X#post.contenuto_post,Y#utenti.nome,Y#utenti.cognome} || X <- mnesia:table(post),
						Y <- mnesia:table(utenti),
						X#post.id_utente =:= Y#utenti.id])),
						QL =qlc:sort(QH,{order,descending}),
						QC = qlc:cursor(QL),
						Numero = (NumeroPagina-1)*20,
						PostNonConsiderati = qlc:next_answers(QC,Numero),
						Post = qlc:next_answers(QC,20),
						qlc:delete_cursor(QC),
						Post
		end.
		
		
%funzione che legge un singolo post
leggi_post(NumeroPost) ->
	do(qlc:q([{X#post.nome_post, X#post.contenuto_post, X#post.data, Y#utenti.nome, Y#utenti.cognome} || X <- mnesia:table(post),
					Y <- mnesia:table(utenti),
					X#post.id_utente =:= Y#utenti.id,
				    X#post.id_post =:= NumeroPost
					])).

%funzione che legge i commenti di un post
leggi_commenti(NumeroPost) ->
	QH = do(qlc:q([{X#commenti.id_commento, X#commenti.commento, X#commenti.data_commento, Y#utenti.nome, Y#utenti.cognome} || X <- mnesia:table(commenti),
						Y <- mnesia:table(utenti),
						X#commenti.id_utente =:= Y#utenti.id,
					    X#commenti.id_post =:= NumeroPost
						])),
	QL = qlc:sort(QH,{order,descending}),
	QC = qlc:cursor(QL),
	Commenti = qlc:next_answers(QC,all_remaining),
	qlc:delete_cursor(QC),
	Commenti.

%funzione che registra un commento all'interno del DB
inserisci_commento(NomeUtente,Commento,Post) ->
			mnesia:dirty_update_counter(appoggio,numCommenti,+1),
			ListaId=leggi_id(numCommenti),
			[Id | Lista ] = ListaId,
			Val = do(qlc:q([{X#utenti.id} || X <- mnesia:table(utenti),
				X#utenti.nome_utente =:= NomeUtente])),
			[{IdUtente}] = Val,
			DataCommento = erlang:localtime(),
			Row = #commenti{id_commento=Id, id_post=Post, id_utente=IdUtente, commento=Commento, data_commento=DataCommento},
   			F = fun() ->
						mnesia:write(Row)
						end,
   			mnesia:transaction(F).

%funzione che restituisce il numero di post totali inseriti all'interno dell'applicazione (per la sezione notizie)
numero_post_totali() ->
		Post = do(qlc:q([X || X <- mnesia:table(post)])),
		Num_Post = length(Post),
		Num_Post.

%funzione che restituisce il numero di post totali inseriti all'interno dell'applicazione (per la sezione notizie)
numero_post_totaliUtente(NomeUtente) ->
		Post = do(qlc:q([{X#post.id_post,X#post.data,X#post.nome_post,X#post.contenuto_post,Y#utenti.nome,Y#utenti.cognome} || X <- mnesia:table(post),
						Y <- mnesia:table(utenti),
						X#post.id_utente =:= Y#utenti.id,
						Y#utenti.nome_utente =:= NomeUtente])),		
		Num_Post = length(Post),
		Num_Post.

%funzione che legge il numero di commenti di un certo post (serve per homePage_utente.yaws, per archivio.yaws e per notizie.yaws)
numero_commenti(Id_post) ->
	Commenti = leggi_commenti(Id_post),
	Num_Commenti = length(Commenti),
	Num_Commenti.


provaQuery() -> 
	do(qlc:q([X || X <- mnesia:table(utenti)])).
	
provaQueryUtenti() -> 
	do(qlc:q([X || X <- mnesia:table(appoggio)])).

provaQueryAvatar() -> 
	do(qlc:q([X || X <- mnesia:table(avatar)])).

provaQueryPost() -> 
	do(qlc:q([X || X <- mnesia:table(post)])).		

provaQueryCommenti() -> 
	do(qlc:q([X || X <- mnesia:table(commenti)])).

	provaQueryChat() -> 
	do(qlc:q([X || X <- mnesia:table(chat)])).	

File: PHP/funzioni.php
<?php
	require_once("config.php");
	//dbConnect: funzione che si connette al database
	function dbConnect(){
		global $config;
		$conn = mysql_connect($config["db_server"], $config["db_utente"], $config["db_password"]) or die ("Errore nella connessione al db: " . mysql_error());
		mysql_select_db($config["db_database"]) or die ("Errore nella selezione del db: " . mysql_error());
		mysql_query("set names 'utf8'");
		return $conn;	
	}
	
		
	//login_utente: verifica che i dati passati come parametri coincidano con quelli presenti nel database
	function login_utente($utente, $password){
		global $config;
		//inizio (o continuazione) di una sessione
		session_start();
		$_SESSION["utente"] = $utente;
		$conn = dbConnect();
		$sql = "SELECT pass FROM utenti WHERE nome_utente = '" . $utente . "'";
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n" . mysql_error());
		if(mysql_num_rows($risposta) == 0)
			return FALSE;
		
		$riga = mysql_fetch_row ($risposta);
		mysql_close($conn);
		
		return ($password == $riga[0]);	
	}
	
		//registra_utente: funzione che restituisce TRUE se l'utente è stato inserito (e quindi registrato) correttamente sul DB, altrimenti restituisce FALSE se l'utente è già presente (e quindi la registrazione non è andata a buon fine)
	function registra_utente($nome, $cognome, $nascita, $nome_utente, $email, $pass ) {
		
		$conn = dbConnect();
		
		$sql = "SELECT nome_utente FROM utenti WHERE nome_utente = '" . $nome_utente . "'";
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n" . mysql_error());
		//se la risposta alla query è diversa da 0 (quindi contiene un utente con lo stesso nome), restituisce falso
		if(mysql_num_rows($risposta) != 0)
			return FALSE;
	
		
		$sql = "INSERT INTO utenti(nome, cognome, nascita, nome_utente, email, pass) VALUES ('" . $nome . "' , '" . $cognome . "', '" . $nascita. "' , '" . $nome_utente . "' , '" . $email . "' , '" . $pass . "')";
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		
		mysql_close ($conn);
		return TRUE;
		
	}
	
	//funzione che legge i post di un singolo utente
		function leggi_post_utente($da, $quanti, $utente){
		//connessione al DB
		$conn = dbConnect();
		$sql = "select post.id_post, post.`data`, post.nome_post, post.contenuto_post, utenti.nome, utenti.cognome from post inner join utenti on post.id_utente = utenti.id where utenti.nome_utente = '" . $utente . "' order by post.`data` DESC LIMIT " . $da . "," . $quanti;
		//prepara la query da passare al DB
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query " . $sql . "\n" . mysql_error());
		$risultato = array();
		while($riga = mysql_fetch_row($risposta))
			$risultato[] = $riga;

			mysql_close($conn);
			return $risultato;
		
	}
	
	//funzione che legge un singolo post
		function leggi_post($id){
		//connessione al DB
		$conn = dbConnect();
		$sql = "select post.id_post, post.`data`, post.nome_post, post.contenuto_post, utenti.nome, utenti.cognome from post inner join utenti on post.id_utente = utenti.id where post.id_post = " . $id;
		//prepara la query da passare al DB
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query " . $sql . "\n" . mysql_error());
		$risultato = array();
		
		$riga = mysql_fetch_row($risposta);

		mysql_close ($conn);		
		return $riga;
		
	}
	
	//funzione che legge i post di tutti gli utenti per la sezione notizie
		function leggi_post_utenti($da, $quanti){
		//connessione al DB
		$conn = dbConnect();
		$sql = "select post.id_post, post.`data`, post.nome_post, post.contenuto_post, utenti.nome, utenti.cognome from post inner join utenti on post.id_utente = utenti.id order by post.`data` DESC LIMIT " . $da . "," . $quanti;
		//prepara la query da passare al DB
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query " . $sql . "\n" . mysql_error());
		$risultato = array();
		while($riga = mysql_fetch_row($risposta))
			$risultato[] = $riga;

			mysql_close($conn);
			return $risultato;
		
	}
	
		//funzione che registra un post di un determinato utente
	function registra_post ($titolo, $testo , $utente){
		//connessione al database
		$conn = dbConnect();
		
		//seleziona il nome e il cognome dell'utente dal DB
		$sql = "select utenti.id from utenti where utenti.nome_utente ='" . $utente . "'";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$riga = mysql_fetch_row ($risposta);
		
		$id = $riga[0];

		//rende conforme il titolo e il testo
		$titolo = nl2br(htmlentities($titolo));
		$testo = nl2br(htmlentities($testo));
		$data = date("Y-m-d H:i:s");
		//crea la query da passare al DB
		$sql = "INSERT INTO post(id_utente, nome_post, contenuto_post, data) VALUES ('" . $id . "' , '" . $titolo . "' , '" . $testo . "' , '" . $data . "')";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		
		mysql_close ($conn);
		
	}
	
	//funzione che legge i commenti di un determinato post
	function leggi_commenti($id){
		//connessione al database
		$conn = dbConnect();
		
		//seleziona il nome e il cognome dell'utente dal DB
		$sql = "select utenti.nome, utenti.cognome, commenti.commento, commenti.data_commento from utenti inner join commenti on utenti.id = commenti.id_utente where commenti.id_post = " . $id . " ORDER BY commenti.data_commento DESC";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$risultato = array();
		while($riga = mysql_fetch_row($risposta))
			$risultato[] = $riga;

			mysql_close($conn);
			return $risultato;
		
	}
	
	//funzione che registra un commento di un utente a un determinato post
	function registra_commento($id_post, $commento , $utente){
		//connessione al database
		$conn = dbConnect();
		
		//seleziona il nome e il cognome dell'utente dal DB
		$sql = "select utenti.id from utenti where utenti.nome_utente ='" . $utente . "'";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$riga = mysql_fetch_row ($risposta);
		
		$id = $riga[0];

		//rende conforme il titolo e il testo
		$commento = nl2br(htmlentities($commento));
		$data = date("Y-m-d H:i:s");
		//crea la query da passare al DB
		$sql = "INSERT INTO commenti(id_post, id_utente, commento, data_commento) VALUES ('" . $id_post . "' , '" . $id . "' , '" . $commento . "' , '" . $data . "')";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		
		mysql_close ($conn);
		
	}
	
	//funzione che recupera il numero commenti di un determinato post
	function numero_commenti($id_post){
		//connessione al database
		$conn = dbConnect();
		
		//seleziona il nome e il cognome dell'utente dal DB
		$sql = "select count(commenti.commento) from commenti where commenti.id_post = " . $id_post;
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$riga = mysql_fetch_row ($risposta);
		
		$id = $riga[0];
		
		mysql_close ($conn);
		return $id;
		
	}
	
		//funzione che recupera il numero commenti di un determinato post
	function leggiNumeroPost($id_post){
		//connessione al database
		$conn = dbConnect();
		
		//seleziona il nome e il cognome dell'utente dal DB
		$sql = "select count(post.id_post) from post";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$riga = mysql_fetch_row ($risposta);
		
		$id = $riga[0];
		
		mysql_close ($conn);
		return $id;
		
	}
	
	//recupera informazioni utente
	function recuperaInformazioni($nome_utente){
		$conn = dbConnect();
		
		$sql = "select * from utenti where utenti.nome_utente ='". $nome_utente ."'";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$riga = mysql_fetch_row ($risposta);

		mysql_close ($conn);
		return $riga;
	}
	
	//verifica se l'utente ha caricato un avatar o meno
	function avatarCaricato($nome_utente){
		$conn = dbConnect();
		
		$sql = "select nome_immagine from avatar inner join utenti on avatar.id_utente = utenti.id where utenti.nome_utente ='". $nome_utente ."'";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		if(mysql_num_rows($risposta) == 0)
			return FALSE;

		mysql_close ($conn);			
		return TRUE;
		

	}
	
	//carica l'avatar dell'utente
	function caricamentoAvatar($nome_utente){
		$conn = dbConnect();
		
		$sql = "select nome_immagine from avatar inner join utenti on avatar.id_utente = utenti.id where utenti.nome_utente ='". $nome_utente ."'";
		//esegue la query
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$riga = mysql_fetch_row ($risposta);

		return "avatar/".$riga[0];
		
		mysql_close ($conn);

	}
	
	//funzione che registra il nuovo avatar dell'utente
	function registraAvatar($nome_utente, $nome_file, $tmp_nome){
		$conn = dbConnect();

		$sql = "select id from utenti where nome_utente = '" . $nome_utente . "'";		
		$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
		$riga = mysql_fetch_row ($risposta);
		
		$id_utente = $riga[0];
		
		if(!avatarCaricato($nome_utente)){
			$sql = "insert into avatar(id_utente, nome_immagine) values ('" . $id_utente . "','" . $nome_file . "')";
			$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
			move_uploaded_file($tmp_nome, "avatar" . "/" . $nome_file);
		} else {
			$conn = dbConnect();
			$sql = "select nome_immagine from avatar where id_utente = '" . $id_utente . "'";		
			$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
			$riga = mysql_fetch_row ($risposta);
			$immagine_vecchia = $riga[0];
			unlink("avatar/" . $immagine_vecchia);
			
			$sql = "update avatar set nome_immagine='" . $nome_file . "' where id_utente=" . $id_utente . "";
			$risposta = mysql_query($sql) or die ("Errore nella query: " . $sql . "\n!" . mysql_error());
			move_uploaded_file($tmp_nome, "avatar/" . $nome_file);
		}
		
		mysql_close ($conn);
		
	}
	
	//verifica se $nomefile ha il tipo di un immagine
	function controllaTipo($nomefile){
		global $tipi_immagine;
		foreach ($tipi_immagine as $tipo)
		//la funzione strrpos parte dalla fine della stringa 
			if (strrpos ($nomefile, $tipo) !== FALSE)
				return TRUE;
		return FALSE;	
	}
?>